Favorite stocks and user profile improvements

kamalq97 opened the pull request:
Changelog:
- Added **profile.html** (for logging out and changing account options) and **watchlist.html** (for viewing and managing favorite stocks) and added paths in the **urls.py** file
- Appended to 'Stock' model in **models.py** class two class methods for adding and removing stock from favorites 
- Defined 'Profile' class model linked one-to-one with user and many-to-many with stock watchlist (favorites)
- Added two methods to 'Profile' class to ensure a profile is automatically created/saved when a user is added/saved
- Under **views.py**, the `index` (homepage) and `single_stock` views (along with their html templates) have been modified by adding favorite management options when a user is logged in (the buttons link to the `watchlist_add` and `watchlist_remove` views)

kamalq97 committed:
Add user profile and favorite stocks

kamalq97 committed:
Delete 0004_profile.py

kamalq97 commented on myapp/views.py:
```diff
@@ -1,23 +1,32 @@
 def single_stock(request, symbol):
 	data = stock_api.get_stock_info(symbol)
-	return render(request, 'single_stock.html', {'page_title': 'Stock Page - %s' % symbol, 'data': data})
+	stock = Stock.objects.get(symbol=symbol)
```
I hope there won't be any merging conflicts in this section of the code, becuase this view function has changed a lot with error handling that @murradkh added

murradkh replied:
It would be for sure, not a problem, we can resolve the conflict. 
for now let's keep it as it is, after we merging into master my/your branch, we fix the conflict to merge it also.

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -2,10 +2,24 @@
+			<td><h2>{{ data.companyName }}<small class="text-muted">({{ data.symbol }}) &emsp14;</small></h2></td>
+			{% if request.user.is_authenticated %}
+			<td>{% if stock in profile.watchlist.all %}
+				<form method="post" action="wremove"> {% csrf_token %}
```
```suggestion
				<form method="post" action={% url 'watchlist_remove' %}> {% csrf_token %}

```

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -2,10 +2,24 @@
+					<button type="submit" class="btn btn-outline-danger"><i class="fa fa-eye-slash"></i> Remove from watchlist</button>
+				</form>
+				{% else %}
+				<form method="post" action="wadd"> {% csrf_token %}
```
the same here
```suggestion
				<form method="post" action={% url 'watchlist_add' %}> {% csrf_token %}

```

kamalq97 replied:
I have fixed it now!

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -2,10 +2,24 @@
+			<td><h2>{{ data.companyName }}<small class="text-muted">({{ data.symbol }}) &emsp14;</small></h2></td>
+			{% if request.user.is_authenticated %}
+			<td>{% if stock in profile.watchlist.all %}
+				<form method="post" action="wremove"> {% csrf_token %}
```
use this, to be more centralized
```suggestion
				<form method="post" action={% url 'watchlist_remove' %}> {% csrf_token %}

```

kamalq97 replied:
Fixed!

murradkh commented on myapp/templates/index.html:
```diff
@@ -29,6 +32,20 @@
                     text-danger
                 {% endif %}
                 ">{{ stock.change_percent|floatformat:2 }}%</td>
+                {% if request.user.is_authenticated %}
```
maybe here you can check if the profile is set or not; you already checked if authenticated in the view.
```suggestion
                {% if profile %}

```

kamalq97 replied:
You are right, I will test this and if there are no problems, I will commit this change

kamalq97 replied:
I am now using `{% if profile %}` everywhere in the templates (expect base.html)

murradkh commented on myapp/views.py:
```diff
@@ -32,12 +41,51 @@ def register(request):
+	profile = Profile.objects.get(id=request.user.id)
+	stock = Stock.objects.filter(symbol=symbol)
+	if not stock.exists():
+		return Response(status=status.HTTP_404_NOT_FOUND)
```
Errors here, no `Response` and `status ` objects imported

kamalq97 replied:
Well spotted! Fixed!

murradkh commented on myapp/urls.py:
```diff
@@ -8,5 +9,9 @@
 	path('accounts/login/', auth_views.LoginView.as_view(template_name='login.html'), name='login'),
 	path('accounts/logout/', views.logout_view, name='logout'),
+	path('stock/watchlist', views.watchlist_view, name='watchlist'),
+	path('stock/<str:symbol>/wremove', views.watchlist_remove_view, name='watchlist_remove'),
```
add the slash as a trailer for the path, for consistency and standardization

kamalq97 replied:
Fixed!

murradkh commented on myapp/templates/index.html:
```diff
@@ -14,6 +14,9 @@
             <th>Name</th>
             <th>Price</th>
             <th>Change</th>
+            {% if request.user.is_authenticated %}
```
this can be shortened to this
```suggestion
            {% if user.is_authenticated %}

```

kamalq97 replied:
I am now using `{% if profile %}` on all templates (except base.html) for consistency

murradkh commented on myapp/templates/single_stock.html:
```diff
@@ -2,10 +2,24 @@
+	<table>
+		<tr>
+			<td><h2>{{ data.companyName }}<small class="text-muted">({{ data.symbol }}) &emsp14;</small></h2></td>
+			{% if request.user.is_authenticated %}
```
the same

murradkh commented on myapp/views.py:
```diff
@@ -1,23 +1,32 @@
 	data = Stock.objects.filter(top_rank__isnull=False).order_by('top_rank')
-	return render(request, 'index.html', {'page_title': 'Main', 'data': data })
+	profile = None
+	if request.user.is_authenticated:
```
..

murradkh commented on myapp/views.py:
```diff
@@ -1,23 +1,32 @@
-	return render(request, 'single_stock.html', {'page_title': 'Stock Page - %s' % symbol, 'data': data})
+	stock = Stock.objects.get(symbol=symbol)
+	profile = None
+	if request.user.is_authenticated:
```
...

kamalq97 replied:
`if user.is_authenticated:` gave me an error 😕
I will leave it as is

murradkh replied:
i meant to change it in the template, my mistake

murradkh commented on myapp/templates/index.html:
```diff
@@ -29,6 +32,20 @@
+                        <button type="submit" class="btn btn-outline-danger" style="padding: 0px 6px;"><i class="fa fa-eye-slash"></i></button>
+                    </form>
+                    {% else %}
+                    <form method="post" action="stock/{{ stock.symbol }}/wadd"> {% csrf_token %}
```
the same for URL, I'm not sure how you can change it, but It could.

kamalq97 replied:
Fixed!

kamalq97 committed:
Update models.py, views.py, and html templates

- Watchlist field in the 'Profile' model now optional to avoid any potential db issues
- Django url tags now used everywhere in the html templates
- Updated profile query in views.py

murradkh reviewed (commented):
looks good in general, organized, I like it; but it's somewhat a big PR, it would be more convenient for the reviewer and for the developer to break it down into small features/PRS.

kamalq97 committed:
Update models.py, views.py, and html templates

- Now using `{% if profile %}` on all html templates  (except for base.html)
- Added CSS on **login.html**
- Fixed issue with adding/removing non-existent stocks in **views.py**
- Paths in the **urls.py** file all have a trailing backslash `/` for more consistency

murradkh reviewed (commented):
the code looks better now, just two comments about the user experience on the changes:
1) when registering users with the same email should popping up an appropriate message.
![Screenshot 2020-10-12 232248](https://user-images.githubusercontent.com/26089554/95787320-f30fb580-0ce1-11eb-9ffd-7bf340406c6b.png)
2) I see that's on each click on the icon "add/remove watch stock", it sends the request and redirects to back to the home page (reloading the whole home page again), which less practical; you can send an AJAX request via javaScript, which sends the request behind the scene. 
but maybe for now to keep it as it is, and make this feature as advanced one to implement later, and to focus now on the essentials.
Good job.

murradkh commented:
when you make the changes notifying my again to approve

kamalq97 commented:
> 1. when registering users with the same email should popping up an appropriate message.

Yeah, I stumbled upon that too, I will create a separate branch and open another pull request to solve that bug 
__


> 2. I see that's on each click on the icon "add/remove watch stock", it sends the request and redirects to back to the home page (reloading the whole home page again), which less practical; you can send an AJAX request via javaScript, which sends the request behind the scene.

I will read more about how to do a POST request in the "background" that does not refresh the page and I will add this in a separate pull request as a nice "enhancement".
__

>    but maybe for now to keep it as it is, and make this feature as advanced one to implement later, and to focus now on the essentials.

Yup! that's the plan 🙂

kamalq97 commented:
> when you make the changes notifying my again to approve

Not sure what you need me to change, I will resolve the registration bug and I will do the non-refresh add/remove favorite in separate pull requests. As you previously said, this pull request is already large and somewhat difficult to review, so I don't want to add more to it.

If there are no issues with the current code changes, maybe we can go ahead and approve this for the time being.

kamalq97 committed:
Merge branch 'master' into StockProject-userprofile

kamalq97 committed:
Update profile.html and views.py

kamalq97 committed:
Merge branch 'master' into StockProject-userprofile

kamalq97 committed:
Update views.py

Re-added accidentally deleted comments

kamalq97 committed:
Update admin.py
